### PR TITLE
Added support for delegations using cloud KMS

### PR DESCRIPTION
**Pull request by kommendorkapten**
This is mainly done by modifying tuf-on-ci-delegate to accept an online key as a signer for delegations. This require some extra care when dealing with the signer's identity. Keyowner or online-uri is now used.

Tuf-on-ci-online-sign now supports an extra parameter --role. This is used to perform online signing with delegations.

Closes https://github.com/theupdateframework/tuf-on-ci/issues/75

### CODE REVIEW

**Comment by kommendorkapten** [comment]
e2e tests are currently failing. Will look into that.

File: signer/tuf_on_ci_sign/delegate.py
Comment (jku) [inline signer/tuf_on_ci_sign/delegate.py:123]: I left some higher level comments on this in the issue: I'm not sure the online keys should be a special case -- why not consider them just one of the signers you can add?

File: repo/tuf_on_ci/online_sign.py
Comment (jku) [inline repo/tuf_on_ci/online_sign.py:33]: I left a higher level comment on the issue: it kind of looks like doing the signing in the repo code is correct but combining the current online-sign and the new feature make this function more complicated but does not have synergies: a new top-level CLI command might make sense

**Review by jku — COMMENTED** [review]
Left some code comments in addition to https://github.com/theupdateframework/tuf-on-ci/issues/75#issuecomment-2257776296

File: signer/tuf_on_ci_sign/_signer_repository.py
Comment (jku) [inline signer/tuf_on_ci_sign/_signer_repository.py:802]: Is this necessary here? We're trying to sign online keys in repo tools not in signer

File: repo/tuf_on_ci/_repository.py
Comment (jku) [reply repo/tuf_on_ci/_repository.py:244]: actually this is not a relevant complaint: this design matches signer and makes sense:
* close() modifies the metadata (versions etc) and may sign as a "side effect"
* sign only signs without any modification

**Comment by jku** [comment]
I have a completely untested branch (based on this one) with an action in https://github.com/jku/tuf-on-ci/tree/external-delegations

I'll have a look at tuf-on-ci-delegate next: I think we should be able to handle online signers just as "one of the signers" and not as a special case (IMO the special casing makes sense for timestamp/snapshot but not really for targets)

File: signer/tuf_on_ci_sign/delegate.py
Comment (jku) [reply signer/tuf_on_ci_sign/delegate.py:123]: I guess my suggestion would require some re-design of adding signers since the KMS keys need more info so maybe not worth it

**Comment by jku** [comment]
I have a seemingly working signing event in https://github.com/jku/test-online-targets/pull/3. Some notes:
* The decision to include in signing event seems good in that signing event is useful to see the status: this should work in external repo and upstream
* when used like this, the signing-event action of course needs similar things as online sign
  * token permission
  * kms account details (e.g. two new optional inputs for GCP)
  * the code to sign into kms
* we could still leave all of those out of the template workflow (and just document hos to add it): if the default signing-event does not need token permission then maybe we don't enable it by default ?

Overall I don't like how the most complicated bits of code got more complicated (in this PR as a whole): testing and refactoring is probably needed... but I think we can work with this

**Comment by jku** [comment]
I've now tested both signing in the upstream project and signing externally. Both work, see the example event.

That said the external flow requires some polishing: 
* I modified workflows compared to upstream
   * I removed most workflows   
   * The signing-event workflow required changes (roles_to_sign and the KMS details)
* These changes mean that the signing event needs to be rebased on top of them when signing externally, and then the signature commit kind of needs to be cherry-picked to the real signing event branch -- because we wouldn't want the workflow modifications getting merged upstream

So the process works, it's just not very ergonomic. Not sure what would be the easiest workflow here

File: signer/tuf_on_ci_sign/delegate.py
Comment (jku) [reply signer/tuf_on_ci_sign/delegate.py:52]: so I am wrong here: it is used (since otherwise we don't have the keyvalue) 

The original design is just bad and this makes it more visible. The key should likely just be a part of the OfflineConfig: `OfflineConfig.signers` is currently a `list[str]` and it probably should be a `list[tuple[str, Key | None]]` so each signer could have a key attached to them

**Comment by kommendorkapten** [comment]
I've tested the updated changes now, and are happy with how they can be used. Will review the code now.

File: signer/tuf_on_ci_sign/delegate.py
Comment (kommendorkapten) [reply signer/tuf_on_ci_sign/delegate.py:95]: The only reason I didn't let `targets` use online keys was to minimize the scope of this PR, i.e not change any existing behaviour. I'm fine to let `targets` use online keys though, I'll update this PR to support it.

Comment (kommendorkapten) [reply signer/tuf_on_ci_sign/delegate.py:95]: This is now fixes.

File: signer/tuf_on_ci_sign/delegate.py
Comment (kommendorkapten) [reply signer/tuf_on_ci_sign/delegate.py:123]: That was the reason I made them special, the code is easier. And also I think the flow for the repository maintainers is simplified if they just provide all the details for the online key direct.

File: signer/tuf_on_ci_sign/_signer_repository.py
Comment (kommendorkapten) [reply signer/tuf_on_ci_sign/_signer_repository.py:802]: It appears it's a left over from an earlier version, good catch. I tried locally with this hunk reverted and it works. Removing this.

File: repo/tuf_on_ci/signing_event.py
Comment (jku) [inline repo/tuf_on_ci/signing_event.py:255]: this kind of also encodes the assumption that there is a single signer: is_signed() returns true if role is signed by threshold of sigs (and is not in signing period) -- I suppose that's fine as long as online signing always means 1 signer

**Review by jku — COMMENTED** [review]
One clearly missing thing is documentation: 
* We should mention how to setup online targets signing: maybe a new `docs/ONLINE_TARGETS_SIGNING.md` or a section in the existing online signing doc
* should document at least the needed changes in signing-event workflow
* I think the external signing part does not need to be explicitly documented
* maybe mention that this is still experimental? 

Otherwise I think we can live with this PR (as long as we file some issues and actually follow up later). I did leave some new code comments for discussion

**Comment by jku** [comment]
btw, for the first time ever I'm now seeing `"You have exceeded a secondary rate limit"` in signing event when testing this (happens calling search api). 

I don't understand how this could be related to the PR but it is the first time I see this -- maybe this limit is per user and I just have enough projects now?

File: actions/signing-event/action.yml
Comment (kommendorkapten) [reply actions/signing-event/action.yml:34]: Oh, good call out, I'll add Azure too.

File: actions/signing-event/action.yml
Comment (kommendorkapten) [reply actions/signing-event/action.yml:115]: But that would be expected, right? First the updated metadata is pushed, then it's signed. Having them as separate commits makes sense. Are you referring to commit twice, but only push once? I don't think that's a big deal.

**Comment by kommendorkapten** [comment]
> btw, for the first time ever I'm now seeing `"You have exceeded a secondary rate limit"` in signing event when testing this (happens calling search api).

I have been seeing this quite frequently, both on my personal test accounts and some real TUF repositories, so probably a coincidence that you started to see it now?

**Comment by jku** [comment]
> I have been seeing this quite frequently, both on my personal test accounts and some real TUF repositories

Well, I filed an issue now that I saw it for the first time

**Comment by jku** [comment]
This is something I was a bit afraid of when modifying signing-event (since it's self-dispatching):

https://github.com/jku/test-online-targets/pull/4

1. there is a signing-event for another role `externally-delegated-role`
1. because `delegated-role` is in signing period, signing-event ends up signing it... not sure we can easily prevent this
1. for some reason it keeps signing again and again -- as if it was constantly in signing period

* So there is the infinite loop to look at -- possibly it has to do with my the long signing period I used (signing period is just a day shorter than expiry)
* but also why is there no signing event for `delegated-role` itself if it is in signing period?

**Comment by kommendorkapten** [comment]
> So there is the infinite loop to look at -- possibly it has to do with my the long signing period I used (signing period is just a day shorter than expiry) 

That is very worrying. I'll take a look.

**Comment by kommendorkapten** [comment]
Somehow there is an event for `delegated-role`: https://github.com/jku/test-online-targets/compare/sign/delegated-role-v2

A bit unexpected but it's commented in the external delegated event: https://github.com/jku/test-online-targets/pull/4#issuecomment-2286970184

**Comment by jku** [comment]
I think it may be just an issue in the date comparison for "are we in signing  period" (in `CIRepository.is_signed()`), I need to take a closer look.

I'll note that this would likely have been noticed if we had a e2e test...

**Comment by jku** [comment]
> Somehow there is an event for `delegated-role`: https://github.com/jku/test-online-targets/compare/sign/delegated-role-v2

oh I know what that is about: the roles are "delegated-role" and "externally-delegated-role": somehow signing event for former finds the PR for the latter because of non-exact string matching

File: actions/online-sign-targets/action.yml
Comment (jku) [inline actions/online-sign-targets/action.yml:103]: we should not have this step or the update-pr step here...
* when using custom token, signing event will trigger and do these steps
* when using default token, we will need to trigger signing-event from here: but that should be safe as it's not a loop then

**Comment by jku** [comment]
### Summary of current approach and its issues 

* We're trying an approach where targets are signed in a new action that is called from a new workflow
* There are multiple ways this should get dispatched (in a normal single repository setup)
   * delegation changes done by maintainer: metadata is modified by user
   * targets files being changes by maintainer: metadata is modified by signing-event
   * create-signing-events bumping the versions: metadata is modified by create-signing-events
* if we use the default token the last two types of changes will not trigger a push event that the sign-online-targets workflow needs

I don't see a reasonable way to make the dispatching work in this design with default token (this is the way signing-event looks like it does otherwise too -- I'd like it to be more granular but it's not possible)


### Options

* Just ditch the action from the PR, admit that all of this code is purely for your use case for now and no-one else will use it: you maintain the action in your repo
* keep (roughly) the current PR. It will not work as is at least with default token (and I don't see a future path where it would)
* revert to signing in signing-event: this we can make work, just need to be careful to avoid dispatch loops

Not great but that's what it looks like :/

**Comment by jku** [comment]
I'm also looking at the loop issue in the "signing integrated in signing-events" version:
* is_signed() is not the correct check in our case:
   * it makes sense in online-sign (timestamp/snapshots case) because then being in signing period actually means we want to sign
   * it does not make sense online-sign-targets  since the role _should_ not be in signing period because the expiry has already been updated: in fact being in signing period is a potential issue: if we sign now there would be a new signing event immediately after merge
   
I have fixed this in my branch and tested in the "infinite loop" signing event.

**Comment by kommendorkapten** [comment]
It now works in this example: https://github.com/kommendorkapten/tuf-repo-test-4/pull/115

There are no loops. Note that I'm using the default token, so the delegation workflow is not [triggered](https://github.com/kommendorkapten/tuf-repo-test-4/blob/main/.github/workflows/new-delegate.yml#L7). I had to manually  approve that workflow to run.

A workaround for this is to not use the default token, but a custom PAT, that would trigger this workflow to run.

I'm fine with this, given that the only _current_ usecase for this is the npm delegation for Sigstore. That will be manged in a separate repo, so the signing workflow will be triggered upon a change, as the push will not happen via the default gh token.

If we accept this, we should create an issue to track this, as for future scenarios, having the complete signing event working is desirable as it would unlock more use-cases for tuf-on-ci.

**Comment by jku** [comment]
Added a commit that makes `is_signed()` actually do what the name says: we are not interested in signing period in the targets signing case (since that is done by create-signing-events)

**Review by jku — APPROVED** [review]

**Comment by jku** [comment]
We can go with this, let's see how it works

**Pull request merged**
